modular/gater/admin_handler.go

```diff
@@ -413,3 +419,223 @@ func (g *GateModular) replicateHandler(w http.ResponseWriter, r *http.Request) {
+	if err != nil {
+		err = ErrInvalidHeader
+		return
+	}
+
+	redundancyIdx := recoveryTask.EcIdx
+	maxRedundancyIndex := int32(params.GetRedundantParityChunkNum()+params.GetRedundantDataChunkNum()) - 1
+	if redundancyIdx < int32(RecoveryMinEcIndex) || redundancyIdx > maxRedundancyIndex {
+		err = executor.ErrRecoveryPieceIndex
+		return
+	}
+
+	var pieceData []byte
+	if redundancyIdx >= 0 {
+		// recovery secondary SP
+		pieceData, err = g.recoverySecondarySP(reqCtx.Context(), chainObjectInfo, bucketInfo, recoveryTask, params, signatureAddr)
+		if err != nil {
+			return
+		}
+	} else {
+		// recovery primary SP
+		pieceData, err = g.recoveryPrimarySP(reqCtx.Context(), chainObjectInfo, bucketInfo, recoveryTask, params, signatureAddr)
+		if err != nil {
+			return
+		}
+	}
+
+	w.Write(pieceData)
+	log.CtxDebugw(reqCtx.Context(), "succeed to get one ec piece data")
+}
+
+func (g *GateModular) recoveryPrimarySP(ctx context.Context, objectInfo *storagetypes.ObjectInfo,
+	bucketInfo *storagetypes.BucketInfo, recoveryTask gfsptask.GfSpRecoveryPieceTask, params *storagetypes.Params, signatureAddr sdktypes.AccAddress) ([]byte, error) {
+	var err error
+	// the primary sp of the object should be consistent with task signature
+	if bucketInfo.PrimarySpAddress != signatureAddr.String() {
+		log.CtxErrorw(ctx, "recovery request not come from primary sp")
+	}
+
+	isOneOfSecondary := false
+	handlerAddr := g.baseApp.OperatorAddress()
+	var ECIndex int
+	for idx, spAddr := range objectInfo.SecondarySpAddresses {
+		if spAddr == handlerAddr {
+			isOneOfSecondary = true
+			ECIndex = idx
+		}
+	}
+	// if the handler SP is not one of the secondary SP of the task object, return err
+	if !isOneOfSecondary {
+		err = ErrRecoverySP
+		return nil, err
+	}
+
+	// init download piece task, get piece data and return the data
+	ECPieceSize := g.baseApp.PieceOp().ECPieceSize(objectInfo.PayloadSize, recoveryTask.GetSegmentIdx(),
+		params.GetMaxSegmentSize(), params.GetRedundantDataChunkNum())
+	ECPieceKey := g.baseApp.PieceOp().ECPieceKey(recoveryTask.GetObjectInfo().Id.Uint64(), recoveryTask.GetSegmentIdx(), uint32(ECIndex))
+
+	pieceTask := &gfsptask.GfSpDownloadPieceTask{}
+	// no need to check quota when recovering primary SP segment data
+	pieceTask.InitDownloadPieceTask(objectInfo, bucketInfo, params, g.baseApp.TaskPriority(pieceTask),
+		false, bucketInfo.PrimarySpAddress, uint64(ECPieceSize), ECPieceKey, 0, uint64(ECPieceSize),
+		g.baseApp.TaskTimeout(pieceTask, uint64(pieceTask.GetSize())), g.baseApp.TaskMaxRetry(pieceTask))
+
+	pieceData, err := g.baseApp.GfSpClient().GetPiece(ctx, pieceTask)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to download piece", "error", err)
+		return nil, downloader.ErrPieceStore
+	}
+
+	return pieceData, nil
+}
+
+func (g *GateModular) recoverySecondarySP(ctx context.Context, objectInfo *storagetypes.ObjectInfo,
```

Comment:
"recoverECPiece" instead of "recoverySecondarySP"

Reply:
fixed